File: code/dart/modular_api/test/handler/usecase_handler_logger_test.dart
import 'dart:convert';
import 'package:test/test.dart';
import 'package:shelf/shelf.dart';
import 'package:modular_api/src/core/schema/field.dart';
import 'package:modular_api/src/core/logger/logger.dart';
import 'package:modular_api/src/core/logger/logging_middleware.dart';
import 'package:modular_api/src/core/usecase/usecase.dart';
import 'package:modular_api/src/core/usecase/usecase_http_handler.dart';
import 'package:modular_api/src/core/usecase/use_case_exception.dart';

class PingInput extends Input {
  final String payload;
  PingInput({required this.payload});

  factory PingInput.fromJson(Map<String, dynamic> json) =>
      PingInput(payload: json['payload'] as String);

  @override
  Map<String, dynamic> toJson() => {'payload': payload};

  @override
  List<SchemaField> get schemaFields =>
      [SchemaField.string('payload', example: 'hello')];

  static PingInput get example => PingInput(payload: 'hello');
}

class PingOutput extends Output {
  final String echo;
  PingOutput({required this.echo});

  @override
  int get statusCode => 200;

  @override
  Map<String, dynamic> toJson() => {'echo': echo};
}

class FailingUseCase implements UseCase<PingInput, PingOutput> {
  @override
  final PingInput input;
  @override
  ModularLogger? logger;
  FailingUseCase(this.input);

  factory FailingUseCase.fromJson(Map<String, dynamic> json) =>
      FailingUseCase(PingInput.fromJson(json));

  @override
  String? validate() => null;

  @override
  Future<PingOutput> execute() async {
    throw UseCaseException(statusCode: 422, message: 'business rule violated');
  }
}

class CrashingUseCase implements UseCase<PingInput, PingOutput> {
  @override
  final PingInput input;
  @override
  ModularLogger? logger;
  CrashingUseCase(this.input);

  factory CrashingUseCase.fromJson(Map<String, dynamic> json) =>
      CrashingUseCase(PingInput.fromJson(json));

  @override
  String? validate() => null;

  @override
  Future<PingOutput> execute() async {
    throw StateError('unexpected null');
  }
}

Handler wrapWithLogging(Handler inner, List<String> logLines) {
  final mw = loggingMiddleware(
    logLevel: LogLevel.debug,
    serviceName: 'test-svc',
    sink: CapturingSink(logLines),
  );
  return mw(inner);
}

class CapturingSink implements StringSink {
  final List<String> lines;
  CapturingSink(this.lines);

  @override
  void write(Object? object) => lines.add(object.toString());
  @override
  void writeAll(Iterable<dynamic> objects, [String separator = '']) =>
      lines.add(objects.join(separator));
  @override
  void writeCharCode(int charCode) => lines.add(String.fromCharCode(charCode));
  @override
  void writeln([Object? object = '']) => lines.add(object.toString());
}

Request postJson(String path, Map<String, dynamic> body, {String? traceId}) {
  final headers = <String, String>{
    'content-type': 'application/json',
    if (traceId != null) 'X-Request-ID': traceId,
  };
  return Request('POST', Uri.parse('http://localhost$path'),
      body: jsonEncode(body), headers: headers);
}

void main() {
  group('useCaseHttpHandler scoped-logger integration (issue #7)', () {
    test('logs UseCaseException through scoped logger with trace_id', () async {
      final logLines = <String>[];
      final handler = wrapWithLogging(
        useCaseHttpHandler(FailingUseCase.fromJson,
            inputExample: PingInput.example),
        logLines,
      );
      const traceId = 'trace-uce-dart-001';
      final response =
          await handler(postJson('/test', {'payload': 'hi'}, traceId: traceId));

      expect(response.statusCode, equals(422));

      final errorLogs = logLines
          .map((l) => jsonDecode(l) as Map<String, dynamic>)
          .where((e) =>
              e['level'] == 'error' &&
              (e['msg'] as String).contains('UseCaseException'))
          .toList();

      expect(errorLogs, isNotEmpty,
          reason: 'expected error log from scoped logger');
      expect(errorLogs.first['trace_id'], equals(traceId));
    });

    test('logs unexpected errors through scoped logger with trace_id',
        () async {
      final logLines = <String>[];
      final handler = wrapWithLogging(
        useCaseHttpHandler(CrashingUseCase.fromJson,
            inputExample: PingInput.example),
        logLines,
      );
      const traceId = 'trace-crash-dart-002';
      final response =
          await handler(postJson('/test', {'payload': 'hi'}, traceId: traceId));

      expect(response.statusCode, equals(500));

      final errorLogs = logLines
          .map((l) => jsonDecode(l) as Map<String, dynamic>)
          .where((e) =>
              e['level'] == 'error' &&
              (e['msg'] as String).contains('Unexpected error'))
          .toList();

      expect(errorLogs, isNotEmpty,
          reason: 'expected error log from scoped logger');
      expect(errorLogs.first['trace_id'], equals(traceId));
    });

    test('does not throw when logger is unavailable', () async {
      final handler = useCaseHttpHandler(FailingUseCase.fromJson,
          inputExample: PingInput.example);

      final response = await handler(postJson('/test', {'payload': 'hi'}));

      expect(response.statusCode, equals(422));
      final body = jsonDecode(await response.readAsString());
      expect(body['message'], equals('business rule violated'));
    });
  });
}

File: code/ts/modular_api/test/handler/usecase_handler_logger.test.ts
// ============================================================
// handler/usecase_handler_logger.test.ts
// Validates that useCaseHandler catch blocks log through the
// request-scoped logger (with trace_id) instead of
// console.error (issue #7).
// ============================================================

import { describe, it, expect } from 'vitest';
import express from 'express';
import request from 'supertest';
import { Input, Output, UseCase, type UseCaseFactory } from '../../src/core/usecase';
import { Field } from '../../src/core/schema/field';
import { useCaseHandler } from '../../src/core/usecase_handler';
import { loggingMiddleware } from '../../src/core/logger/logging_middleware';
import { LogLevel } from '../../src/core/logger/logger';
import { UseCaseException } from '../../src/core/use_case_exception';

// ── Minimal DTOs ───────────────────────────────────────────

class PingInput extends Input {
  @Field.string({ description: 'payload' })
  payload!: string;

  static fromJson(json: Record<string, unknown>): PingInput {
    const i = new PingInput();
    i.payload = json['payload'] as string;
    return i;
  }
}

class PingOutput extends Output {
  @Field.string({ description: 'echo' })
  echo!: string;
  get statusCode() { return 200; }
}

// ── UseCase that throws UseCaseException ───────────────────

class FailingUseCase extends UseCase<PingInput, PingOutput> {
  readonly input: PingInput;
  constructor(input: PingInput) { super(); this.input = input; }

  static fromJson(json: Record<string, unknown>): FailingUseCase {
    return new FailingUseCase(PingInput.fromJson(json));
  }

  validate() { return null; }

  async execute(): Promise<PingOutput> {
    throw new UseCaseException({ statusCode: 422, message: 'business rule violated' });
  }
}

// ── UseCase that throws unexpected error ───────────────────

class CrashingUseCase extends UseCase<PingInput, PingOutput> {
  readonly input: PingInput;
  constructor(input: PingInput) { super(); this.input = input; }

  static fromJson(json: Record<string, unknown>): CrashingUseCase {
    return new CrashingUseCase(PingInput.fromJson(json));
  }

  validate() { return null; }

  async execute(): Promise<PingOutput> {
    throw new TypeError('cannot read property x of undefined');
  }
}

// ── Helper ─────────────────────────────────────────────────

function buildApp(
  factory: UseCaseFactory,
  logLines: string[],
  inputClass?: abstract new (...args: unknown[]) => Input,
) {
  const app = express();

  app.use(
    loggingMiddleware({
      logLevel: LogLevel.debug,
      serviceName: 'test-svc',
      writeFn: (line: string) => logLines.push(line),
    }),
  );

  app.use(express.json());
  app.post('/test', useCaseHandler(factory, { inputClass }));
  return app;
}

// ── Tests ──────────────────────────────────────────────────

describe('useCaseHandler scoped-logger integration (issue #7)', () => {
  it('logs UseCaseException through scoped logger with trace_id', async () => {
    const logLines: string[] = [];
    const app = buildApp(FailingUseCase.fromJson as UseCaseFactory, logLines, PingInput);
    const traceId = 'trace-uce-001';

    const res = await request(app)
      .post('/test')
      .set('X-Request-ID', traceId)
      .send({ payload: 'hello' });

    expect(res.status).toBe(422);

    const errorLog = logLines
      .map((l) => JSON.parse(l))
      .find(
        (e: Record<string, unknown>) =>
          e['level'] === 'error' && (e['msg'] as string).includes('UseCaseException'),
      );

    expect(errorLog).toBeDefined();
    expect(errorLog['trace_id']).toBe(traceId);
  });

  it('logs unexpected errors through scoped logger with trace_id', async () => {
    const logLines: string[] = [];
    const app = buildApp(CrashingUseCase.fromJson as UseCaseFactory, logLines, PingInput);
    const traceId = 'trace-crash-002';

    const res = await request(app)
      .post('/test')
      .set('X-Request-ID', traceId)
      .send({ payload: 'hello' });

    expect(res.status).toBe(500);

    const errorLog = logLines
      .map((l) => JSON.parse(l))
      .find(
        (e: Record<string, unknown>) =>
          e['level'] === 'error' &&
          (e['msg'] as string).includes('Unexpected error'),
      );

    expect(errorLog).toBeDefined();
    expect(errorLog['trace_id']).toBe(traceId);
  });

  it('does not throw when logger is unavailable (excluded routes)', async () => {
    // Build an app WITHOUT loggingMiddleware — simulates excluded route
    const app = express();
    app.use(express.json());
    app.post('/test', useCaseHandler(FailingUseCase.fromJson as UseCaseFactory, { inputClass: PingInput }));

    const res = await request(app)
      .post('/test')
      .send({ payload: 'hello' });

    // Should still return the correct error response even without logger
    expect(res.status).toBe(422);
    expect(res.body).toHaveProperty('message', 'business rule violated');
  });
});
